import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/backupuploader/BackupUploadService.java

```
@@ -0,0 +1,112 @@
+    if (backupState.decrement()) {
+      int attemptNumber = attemptsAmount - backupState.getAttemptsCountdown();
+
+      log.info("Trying to resend data, sopInstanceUID={}, attempt № {}. ", fileName, attemptNumber);
+      CompletableFuture completableFuture =
+          CompletableFuture.supplyAsync(
+              () -> {
+                try (ByteArrayInputStream bais = new ByteArrayInputStream(bytes)) {
+                  webClient.stowRs(bais);
+                  removeBackup(backupState);
+                  log.debug(
+                      "sopInstanceUID={}, bytes={},\n resend attempt № {}, - successful.", fileName, bytes, attemptNumber);
+                } catch (IOException | IDicomWebClient.DicomWebException ex) {
+                  log.error("sopInstanceUID={}, resend attempt № {} - failed.",
+                      fileName, attemptsAmount - backupState.getAttemptsCountdown(), ex);
+                  throw new CompletionException(ex);
+                }
+                return null;
+              },
+              CompletableFuture.delayedExecutor(
+                  delayCalculator.getExponentialDelayMillis(backupState.getAttemptsCountdown()),
+                  TimeUnit.MILLISECONDS))
+          .exceptionally(
+              ex -> {
+                if (ex.getCause() instanceof IDicomWebClient.DicomWebException) {
+                  if (backupState.getAttemptsCountdown() > 0) {
+                    scheduleUploadWithDelay(webClient, bytes, backupState);
+                  } else {
+                    log.debug("sopInstanceUID={}, No resend attempt left.", fileName);
+                  }
+                } else {
+                  MonitoringService.addEvent(Event.CSTORE_BACKUP_ERROR);
+                  log.error("sopInstanceUID={}, read backup failed.", fileName, ex.getCause());
+                }
+                return null;
+              });
+    } else {
+      log.info("sopInstanceUID={}, Backup resend attempts exhausted.", fileName);
```

We should throw an exception here so the status in cstore is written as failed